Refactor dynamic library error checking on *nix

ecstatic-morse opened the pull request:
The old code was checking `dlerror` more often than necessary, since (unlike `dlsym`) checking the return value of [`dlopen`](https://www.man7.org/linux/man-pages/man3/dlopen.3.html) is enough to indicate whether an error occurred. In the first commit, I've refactored the code to minimize the number of system calls needed. It should be strictly better than the old version.

The second commit is an optional addendum which fixes the issue observed on illumos in #74469, a PR I reviewed that was ultimately closed due to inactivity. I'm not sure how hard we try to work around platform-specific bugs like this, and I believe that, due to the way that `dlerror` is specified in the POSIX standard, libc implementations that want to run on conforming systems cannot call `dlsym` in multi-threaded programs.

ecstatic-morse committed:
Refactor dynamic library error checking on *nix

The old code was checking `dlerror` more often than necessary, since the
return value of `dlopen` indicates whether an error occurred.

ecstatic-morse committed:
Treat a NULL return from `dlsym` as an error on illumos

This works around behavior observed on illumos in #74469, in which
foreign code (libc according to the OP) was racing with rustc to check
`dlerror`.

ollie27 commented on src/librustc_metadata/dynamic_lib.rs:
```diff
@@ -51,51 +51,97 @@ mod tests;
+        // Because symbols with address zero shouldn't occur in practice, we
+        // treat them as errors on platforms with misbehaving libc
+        // implementations.
+        const DLSYM_NULL_IS_ERROR: bool = cfg!(target_os = "illumos");
```
It's probably safer to always treat null as an error. It doesn't look as if any current uses of `symbol` actually check for null so if this did ever return null then we'd have UB.

nagisa commented on src/librustc_metadata/dynamic_lib.rs:
```diff
@@ -51,51 +51,97 @@ mod tests;
-            libc::dlopen(s.as_ptr(), libc::RTLD_LAZY) as *mut u8
-        })
-    }
+    // `dlerror` is process global, so we can only allow a single thread at a
```
Nit: `dlerror` It is MT-safe (thread local) on most modern POSIX implementations. Exceptions include freebsd and dragonflybsd as per investigation I did early this year, and apparently illumos as I’ve just learnt.

POSIX in particular allows for `dlerror` to be MT-safe (since a comparatively recent standard revision), but does not require it.

nagisa commented on src/librustc_metadata/dynamic_lib.rs:
```diff
@@ -51,51 +51,97 @@ mod tests;
+        let s = CString::new(filename.as_bytes()).unwrap();
+
+        let mut dlerror = error::lock();
+        let ret = unsafe { libc::dlopen(s.as_ptr(), libc::RTLD_LAZY) } as *mut u8;
```
nit preexisting, but should also explicitly specify `RTLD_LOCAL`, not specifying `LOCAL` vs `GLOBAL` will end up resulting in platform choosing one for you, and they pick different defaults and may result in confusing bugs!

nagisa commented:
r? @nagisa I guess I’m the most appropriate reviewer for this, given that I maintain a similar crate.

I broadly agree with @ollie27 in that we should just consider `dlsym() == nullptr` to be an error universally. It is not like the implementation here needs to cater for all of the obscure use-cases. This way we can cut some corners and simplify the code significantly (most notably by removing most of the motivation to use the `Mutex`).

ecstatic-morse committed:
Always treat `dlsym` returning NULL as an error

This simplifies the code somewhat. Also updates comments to reflect
notes from reviw about thread-safety of `dlerror`.

ecstatic-morse committed:
Explicitly pass `RTLD_LOCAL` to `dlopen`

This happens to be the default on Linux, but the default is unspecified
in the POSIX standard. Also switches to `cast` to keep line lengths in
check.

ecstatic-morse commented:
The most recent changes treat a `NULL` from `dlsym` as an error. I also expanded the comments to reflect the 2017 revision to the POSIX standard that @nagisa alluded to above and started passing `RTLD_LOCAL` to `dlopen`.

I don't feel comfortable removing the locking entirely in this PR, since it's not a performance issue and can give more predictable error messages on lower tier platforms that don't have thread-safe`dlerror`. You're free to remove the `error` module in a later PR if you think it's still too complex, but I think this is better than the status quo.

nagisa commented:
@bors r+

Yeah LGTM.

bors commented:
:pushpin: Commit aae6c0fbfe3051e5539f47d0e9d84ddee53f72bd has been approved by `nagisa`

<!-- @bors r=nagisa aae6c0fbfe3051e5539f47d0e9d84ddee53f72bd -->
<!-- homu: {"type":"Approved","sha":"aae6c0fbfe3051e5539f47d0e9d84ddee53f72bd","approver":"nagisa"} -->

bors commented:
:hourglass: Testing commit aae6c0fbfe3051e5539f47d0e9d84ddee53f72bd with merge 3e98860425447e30e967e83a738f7c54b7fe8273...
<!-- homu: {"type":"BuildStarted","head_sha":"aae6c0fbfe3051e5539f47d0e9d84ddee53f72bd","merge_sha":"3e98860425447e30e967e83a738f7c54b7fe8273"} -->

jclulow reviewed (approved):
This is great, and definitely seems like it will cover the thing I submitted an unpolished patch for earlier.  Thank you!

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1029572259), [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=36226)
Approved by: nagisa
Pushing 3e98860425447e30e967e83a738f7c54b7fe8273 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"nagisa","base_ref":"master","builders":{"checks-azure":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=36226","checks-actions":"https://github.com/rust-lang-ci/rust/runs/1029572259"},"merge_sha":"3e98860425447e30e967e83a738f7c54b7fe8273"} -->
